packages/lib/test/computedTree/computedTree.test.ts

```diff
@@ -0,0 +1,518 @@
+import { computed, reaction } from "mobx"
+import {
+  computedTree,
+  createContext,
+  detach,
+  findChildren,
+  findParent,
+  findParentPath,
+  getChildrenObjects,
+  getParent,
+  getParentPath,
+  getParentToChildPath,
+  getRefsResolvingTo,
+  getRoot,
+  getRootPath,
+  getSnapshot,
+  idProp,
+  isArray,
+  isChildOfParent,
+  isComputedTreeNode,
+  isParentOfChild,
+  isPlainObject,
+  isRoot,
+  isTreeNode,
+  model,
+  Model,
+  onChildAttachedTo,
+  onPatches,
+  prop,
+  Ref,
+  registerRootStore,
+  resolvePath,
+  rootRef,
+  unregisterRootStore,
+  walkTree,
+  WalkTreeMode,
+} from "../../src"
+import "../commonSetup"
+import { autoDispose } from "../utils"
+
+@model("R")
+class R extends Model({
+  id: idProp,
+  m: prop<M | undefined>().withSetter(),
+  mRef: prop<Ref<M> | undefined>().withSetter(),
+  value: prop<number>(0).withSetter(),
+}) {
+  @computedTree
+  get string() {
+    return "abc"
+  }
+
+  @computedTree
+  get number() {
+    return 42
+  }
+
+  @computedTree
+  get boolean() {
+    return true
+  }
+
+  @computedTree
+  get null() {
+    return null
+  }
+
+  @computedTree
+  get undefined() {
+    return undefined
+  }
+
+  @computedTree
+  get array() {
+    return ["abc"]
+  }
+
+  @computedTree
+  get plainObject() {
+    return { key: "value" }
+  }
+
+  @computedTree
+  get model() {
+    return new M({ id: `${this.id}.model`, value: this.value })
+  }
+
+  @computedTree
+  get ref() {
+    return this.m ? createMRef(this.m) : undefined
+  }
+
+  @computedTree
+  get modelAlreadyAttached() {
+    return this.m
+  }
+
+  @computedTree
+  get modelIsParent() {
+    return this
+  }
+}
+
+@model("M")
+class M extends Model({
+  id: idProp,
+  value: prop<number>(1).withSetter(),
+}) {
+  isAttachedToRootStore?: boolean
+
+  onAttachedToRootStore() {
+    this.isAttachedToRootStore = true
+    return () => {
+      this.isAttachedToRootStore = false
+    }
+  }
+
+  @computed
+  get backrefs() {
+    return getRefsResolvingTo(this)
+  }
+
+  @computed
+  get isComputedTreeNode() {
+    return isComputedTreeNode(this)
+  }
+}
+
+const createMRef = rootRef<M>("MRef")
+
+test("computed tree decorator cannot decorate a 'get' accessor of a plain class", () => {
+  expect(() => {
+    // @ts-ignore
+    // eslint-disable-next-line
+    class X {
+      @computedTree
+      get value() {
+        return null
+      }
+    }
+  }).toThrow("@computedTree can only decorate 'get' accessors of class or data models")
+})
+
+test("computed tree decorator cannot decorate a 'set' accessor", () => {
+  expect(() => {
+    // @ts-ignore
+    // eslint-disable-next-line
+    class X extends Model({}) {
+      @computedTree
+      set value(_value: any) {}
+    }
+  }).toThrow("@computedTree requires a 'get' accessor")
+})
+
```

Comment:
(optional) cannot decorate a method / cannot decorate a field

Reply:
Are you suggesting to change the error message or add tests that check that the decorator cannot decorate a method or field?

Reply:
The last

Reply:
I've added a test that checks decorating a method. When decorating a field, TS complains already:

> Unable to resolve signature of property decorator when called as an expression. ts(1240)

Would you like a test for that case anyway? For fields, the descriptor is `undefined`, so I'd probably need to make it optional check this case, although when applied to an accessor the descriptor would never be `undefined`, i.e. I'd have to change the typing for a case that TS would disallow anyway.

Reply:
nah, good enough